Automate generation of CiliumNode, CiliumIdentity, & CiliumEndpoint CRDs using controller-gen

aanm opened the pull request:
See https://github.com/cilium/cilium/pull/11476#issuecomment-631162957 for what this PR is about.

Followups are tracked here: https://github.com/cilium/cilium/issues/12737

Updates: https://github.com/cilium/cilium/issues/11142
Related: https://github.com/cilium/cilium/issues/12516

---

Previous PR body
---

> 1/2
> 
> With a CRD generator we will generate the CRD automatically based on specific markers across the golang source code.
> 
> TODO:
> - [x] Create CRD for CCNP (it seems it's only creating for CNP so far)
> - [x] Double check if the validation for all fields are correct
>   - [x] CN
>   - [x] CNP
> - [x] Check if all fields of a CRD are generated. It seems `shortnames` are not present in the generated CRD: 
> ```
>     shortNames:
>     - cnp
>     - ciliumnp
> ```
> - [x] Ensure additional columns is also in the generated CRD yaml 
> ```
> additionalPrinterColumns:
>  - JSONPath: .status.id
>    description: Cilium endpoint id
>    name: Endpoint ID
>    type: integer
>  - JSONPath: .status.identity.id
>    description: Cilium identity id
>    name: Identity ID
>    type: integer
>  - JSONPath: .status.policy.ingress.enforcing
>    description: Ingress enforcement in the endpoint
>    name: Ingress Enforcement
>    type: boolean
>  - JSONPath: .status.policy.egress.enforcing
>    description: Egress enforcement in the endpoint
>    name: Egress Enforcement
>    type: boolean
>  - JSONPath: .status.state
>    description: Endpoint current state
>    name: Endpoint State
>    type: string
>  - JSONPath: .status.networking.addressing[0].ipv4
>    description: Endpoint IPv4 address
>    name: IPv4
>    type: string
>  - JSONPath: .status.networking.addressing[0].ipv6
>    description: Endpoint IPv6 address
>    name: IPv6
>    type: string
>  conversion:
>    strategy: None
> ```
> - [x] Tidy up the creation of the CRDs into golang (currently uses go-bindata, find another solution if avoidable the usage of go-bindata)
> - [ ] ~~Move / Copy the CRDs from `examples/crds` to `install/kubernetes/cilium/crds` similar to what is done by https://github.com/cilium/cilium/pull/11261.~~ Marked as a followup in https://github.com/cilium/cilium/issues/12737.
> 
> 
> Follow ups:
> - [ ] ~~Currently the maintenance of a CRD is performed by cilium-agents, we can move this to the cilium-operator. Cilium-Operator will double check if the installed CRD is up-to-date (this is because helm does not provide a mechanism to upgrade CRDs). Currently the up-to-date check of the CRD is being done by Cilium agent upon initialization. Moving this to the Cilium-Operator will likely decrease the size of the Cilium-agent binary because the bindata for the generated CRD will be built into the CIlium Operator binary.~~
> 
> Resources:
> https://book.kubebuilder.io/reference/generating-crd.html
> https://book.kubebuilder.io/reference/markers/crd-validation.html

christarazi committed:
k8s: Generate CiliumEndpoint CRD

This commit also adds a missing endpoint state, "invalid"

This CRD is generated using
https://book.kubebuilder.io/reference/generating-crd.html.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Generate CiliumIdentity CRD

This CRD is generated using
https://book.kubebuilder.io/reference/generating-crd.html.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Generate CiliumNode CRD

This commit adds resource name generation via markers. It also updates
the validation schema via markers which the old method of hardcoding had
a few fields missing / incorrect.

Here's a summary of the main differences in the validation schema:

1) Old schema had eniTypes.ENISpec.MaxAllocate, which doesn't exist. It
   exists under ipamTypes.IPAMSpec.MaxAllocate.
2) Old schema was missing eniTypes.ENISpec.{InstanceID,InstanceType},
   which is now included in this new generated schema.
3) Old schema was missing EncryptionSpec, which is now included in this
   new generated schema.
4) Old schema has azureTypes.AzureSpec.InstanceID, but that was removed
   and deprecated in favor of ipamTypes.IPAMSpec.InstanceID. See PR
   which forgot to remove this from the old validation schema. [1]

This CRD is generated using
https://book.kubebuilder.io/reference/generating-crd.html.

[1]: https://github.com/cilium/cilium/pull/10569

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Improve readability of createUpdateCRD()

Kubernetes code tends to have very long lines because of package /
variable names, and can make certain lines hard to parse. This commit
attempts to make the code easier to parse. No functional changes were
made.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Relax CRD upgrade check

Previously, we only wanted to update a CRD when the CRD has an
out-of-date schema version (which is inside the label
"io.cilium.k8s.crd.schema.version".)

Some CRDs like CiliumNode (CN) had a validation schema, but did not have
a schema version (label mentioned above). This commit will allow all
CRDs like CN to be updated, because it removes the check for the
existance of the label.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Bump schema version to 1.22.2

This bump is after generating the following CRDs using the
controller-gen tool:

- CiliumNode
- CiliumIdentity
- CiliumEndpoint

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Restore removed CiliumNode CRD validation

This was accidently removed in commit
225526078b05c0b8d7d37d91cae8a62d45c0a448.

Fixes: 2255260 ("k8s, policy: Add automated CRD generation")

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on pkg/aws/eni/types/types.go:
```diff
@@ -66,33 +66,33 @@ type ENISpec struct {
 	// used for IP allocation, eth0 will be ignored for PodIP allocation.
 	//
-	// +optional
+	// +kubebuilder:validation:Minimum=0
```
This value (and the ones that you added the markers) are optional, will the validation complain if the values are not set? (not set  != "0")

sayboras replied:
as the type is *int, should it be just optional ? Example from kubebuilder repo https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_types.go#L92

christarazi replied:
It shouldn't complain because this validation is not new, see here: https://github.com/cilium/cilium/pull/11476/commits/90cc78f194799a42a4708c01932820de04dad60b#diff-da7eac65a6ed1ef87f8fec512cb90529R330-R333

christarazi replied:
I can mark the types as explicitly optional so that it is clear.

aanm reviewed (commented):
I can't approve my own PR :cry: 

Overall looks good, only a small comment

sayboras commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -214,10 +218,8 @@ type EndpointIdentity struct {
 // cilium-operator uses the list of nodes in status to reference count
 // users of this identity, and to expire stale usage.
 type CiliumIdentity struct {
-	// +k8s:openapi-gen=false
```
might need do similar clean up for `CiliumNode` struct as well.

christarazi replied:
I think this was on accident. But it seems like removing them had no effect, though? I'll remove them all from the CRD types and see what happens.

sayboras reviewed (approved):
One small question only, not a blocker :+1:

christarazi committed:
k8s: Remove unneeded markers from CNP and CCNP

These marker comments have no effect. They are likely leftovers from
when the tooling was not as mature. We are able to generate (`make
generate-k8s-api`) without issue.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

fristonio reviewed (approved):
Awesome :100:  :100:

aanm reviewed (commented):
I can't approve my own PR but LGTM :+1:
